credential/exchange/submission.go

```diff
@@ -0,0 +1,318 @@
+func BuildPresentationSubmissionVP(def PresentationDefinition, claims []PresentationClaim) (*credential.VerifiablePresentation, error) {
+	if err := canProcessDefinition(def); err != nil {
+		return nil, errors.Wrap(err, "feature not supported in processing given presentation definition")
+	}
+	builder := credential.NewVerifiablePresentationBuilder()
+	if err := builder.AddContext(PresentationSubmissionContext); err != nil {
+		return nil, err
+	}
+	if err := builder.AddType(PresentationSubmissionType); err != nil {
+		return nil, err
+	}
+
+	submission := PresentationSubmission{
+		ID:           uuid.New().String(),
+		DefinitionID: def.ID,
+	}
+
+	// begin to process to presentation definition against the available claims
+	var processedClaims []processedClaim
+	for i, id := range def.InputDescriptors {
+		processedInputDescriptor, err := processInputDescriptor(id, claims)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error processing input descriptor: %s", id.ID)
+		}
+		if processedInputDescriptor == nil {
+			return nil, fmt.Errorf("input descrpitor<%s> could not be fulfilled; could not build a valid presentation submission", id.ID)
+		}
+		processedClaims = append(processedClaims, processedClaim{
+			Claim: processedInputDescriptor.Claim,
+			SubmissionDescriptor: SubmissionDescriptor{
+				ID:     processedInputDescriptor.ID,
+				Format: processedInputDescriptor.Format,
+				Path:   fmt.Sprintf("$.verifiableCredential[%d]", i),
+			},
+		})
+	}
+
+	// set descriptor map in submission and credentials to the VP
+	var descriptorMap []SubmissionDescriptor
+	for _, claim := range processedClaims {
+		descriptorMap = append(descriptorMap, claim.SubmissionDescriptor)
+		if err := builder.AddVerifiableCredentials(claim.Claim); err != nil {
+			return nil, errors.Wrap(err, "could not add claim value to verifiable presentation")
+		}
+	}
+
+	// set submission in vp, build, and return
+	if err := builder.SetPresentationSubmission(submission); err != nil {
+		return nil, err
+	}
+	return builder.Build()
+}
+
+// processedInputDescriptor
+type processedInputDescriptor struct {
+	// input descriptor id
+	ID string
+	// generic claim
+	Claim map[string]interface{}
+	// claim format
+	Format string
+}
+
+// limitedInputDescriptor is the claim data after being filtered/limited via JSON path
+type limitedInputDescriptor struct {
+	Path string
+	Data interface{}
+}
+
+// processInputDescriptor runs the input evaluation algorithm described in the spec for a specific input descriptor
+// https://identity.foundation/presentation-exchange/#input-evaluation
+// TODO(gabe) consider normalization of claims before processing
+func processInputDescriptor(id InputDescriptor, claims []PresentationClaim) (*processedInputDescriptor, error) {
+	constraints := id.Constraints
+	fields := constraints.Fields
+	if !(constraints == nil || len(fields) == 0) {
+		return nil, fmt.Errorf("invalid input descriptor without constraints and/or fields: %s", id.ID)
+	}
```

Comment:
Is this logic backwards? i'm reading it as: if constraints isn't nil or there are fields, throw an error saying there aren't constraints or fields

Reply:
<img width="1079" alt="image" src="https://user-images.githubusercontent.com/4887440/159969503-8c98b9cc-2d51-4aef-95e7-823f5c4d5291.png">

given that `constraints` isn't a required property, does it seem correct to return an error? or does it make more sense to return an empty `processedInputDescriptor`?

Reply:
is this correct?
```
if constraints != nil && len(fields) != 0 {
```

Reply:
@decentralgabe i think it'd be:
```go
if (constraints == nil || len(fields) == 0) {
    return nil, fmt.Errorf("invalid input descriptor without constraints and/or fields: %s", id.ID)
}
```

which reads as: if constraints is nil or there are no fields in constraints then there's nothing to do so throw err.

could do it the way you suggested but then the rest of the logic in the function would be nested in that conditional

Reply:
d'oh. thanks

Reply:
```
	constraints := id.Constraints
	if constraints == nil {
		return nil, fmt.Errorf("unable to process input descriptor without constraints")
	}
	fields := constraints.Fields
	if len(fields) != 0 {
		return nil, fmt.Errorf("unable to process input descriptor without fields: %s", id.ID)
	}
```

Reply:
awesome. I was actually curious about if `constraints.Fields` would panic if `constraints` was nil, so i tried it in go playground and the program executed successfully so i decided not to comment about it. The way you have it now reads much cleaner IMO